sympy/polys/rootoftools.py

```diff
@@ -711,94 +716,118 @@ def _n(self, prec):
+        The `eval_approx` method will also find the root to a given
+        precision but the interval is not modified unless the search
+        for the root fails to converge within the root bounds.
+        """
+        # all kwargs are ignored
+        return self.eval_rational(n=prec_to_dps(prec))._evalf(prec)
 
-        The following example first obtains Rational approximation to 1e-7
-        accuracy for all roots of the 4-th order Legendre polynomial, and then
-        evaluates it to 5 decimal digits (so all digits will be correct
-        including rounding):
+    def eval_rational(self, dx=None, dy=None, n=15):
+        """
+        Return a Rational approximation of ``self`` that has real
+        and imaginary component approximations that are within ``dx``
+        and ``dy`` of the true values, respectively. Alternatively,
+        ``n`` digits of precision can be specified.
+
+        The interval is refined with bisection and is sure to
+        converge. The root bounds are updated when the refinement
+        is complete so recalculation at the same or lesser precision
+        will not have to repeat the refinement and should be much
+        faster.
+
+        The following example first obtains Rational approximation to
+        1e-6 accuracy for all roots of the 4-th order Legendre
+        polynomial. Since the roots are all less than 1, this will
+        ensure the decimal representation of the approximation will be
+        correct (including rounding) to 5 digits:
 
         >>> from sympy import S, legendre_poly, Symbol
         >>> x = Symbol("x")
         >>> p = legendre_poly(4, x, polys=True)
-        >>> roots = [r.eval_rational(S(1)/10**7) for r in p.real_roots()]
-        >>> roots = [str(r.n(5)) for r in roots]
-        >>> roots
-        ['-0.86114', '-0.33998', '0.33998', '0.86114']
+        >>> r = p.real_roots()[-1]
+        >>> r.eval_rational(10**-6).n(5)
+        0.86114
 
-        """
+        It is not necessary to a two-step calculation, however: the
+        decimal representation can be computed directly:
 
-        if not (self.is_real or self.is_imaginary):
-            raise NotImplementedError(
-                "eval_rational() only works for real polynomials so far")
-        g = self.poly.gen
-        if not g.is_Symbol:
-            d = Dummy('x')
-            if self.is_imaginary:
-                d *= I
-            func = lambdify(d, self.expr.subs(g, d))
+        >>> r.evalf(17)
+        0.86113631159405258
+
+        """
+        dy = dy or dx
+        if dx:
+            rtol = None
+            dx = dx if isinstance(dx, Rational) else Rational(str(dx))
+            dy = dy if isinstance(dy, Rational) else Rational(str(dy))
         else:
-            expr = self.expr
-            if self.is_imaginary:
-                expr = self.expr.subs(g, I*g)
-            func = lambdify(g, expr)
+            rtol = S(10)**-(n + 1)  # +1 for guard digit
```

Comment:
I'm not so sure the thinking about the guard digit is sound. Consider the number 0.339981043584856xx48. Whether xx -- the 16th and 17th decimal place -- will remain xx or be rounded up to x(x+1) (e.g. 13 -> 14) depends on what xx is and what it is that follows xx. When 47 follows xx, none fail. When 48 follows xx the following values for xx fail:
```
>>> for j in range(100):
...  i = S('0.339981043584856%02d48'%j).n(17)
...  if str(i)[-2:]!=str('%02d'%j):print(j)
...
13
30
47
99
```
Removing the first decimal digit so xx fall at positions 15 and 16 give a different set of xx numbers that fail when trying to evaluate at the last xx digit (in this case, 16):
```
>>> for j in range(100):
...  i = S('0.39981043584856%02d48'%j).n(16)
...  if str(i)[-2:]!=str('%02d'%j):print(j)
...
1
6
11
16
18
23
28
33
40
45
50
57
62
67
74
79
84
91
96
```
It seems like the only way to make sure the last digit is correct is to keep evaluating 1 more decimal place until there is no change.

Reply:
5 binary guard digits at most are needed to code each decimal digit, so we need at most n+2 extra decimal digits to ensure that the nth digit is correct. e.g. prec_to_dps(dps_to_prec(10)+5) -> 12 so 12 digits will be needed to capture the 10th digit correctly.